lib/Runtime/Library/amd64/JavascriptFunctionA.S

```diff
@@ -37,112 +51,94 @@ C_FUNC(amd64_CallFunction):
         // the stack)
         sub  rsp, 8
 
-        mov rbp, rsp
-        
         // We need to call the target function with the following:
         //  rdi = function
         //  rsi = CallInfo
-        //  rdx = argv[0]
-        //  rcx = argv[1]
-        //  r8  = argv[2]
-        //  r9  = argv[3]
-        //  Remaining args on stack
+        //  function, callInfo, and all args on stack
         //
-        // Coming into the function, rdi = function, rsi = entrypoint, 
+        // Coming into the function, rdi = function, rsi = entrypoint,
         // rdx = callInfo, rcx = argc, r8 = argv
         //
         // Lets store the entry point in r11 and set up its first two
         // arguments
         // (Note that since rdi is already set up, it doesn't need to
         // change)
-        
+
         mov r11, rsi // Store entry point in r11
         mov rsi, rdx // Store CallInfo in rs1
-        mov r12, r8  // Store argv in r12
-
-        // Now check how many arguments we have
-        cmp rcx, 5
-        jge copy_args_to_stack
-        cmp rcx, 4
-        je store_arg4
-        cmp rcx, 3
-        je store_arg3
-        cmp rcx, 2
-        je store_arg2
-        cmp rcx, 1
-        je store_arg1
-        jmp args_setup_done
 
-        // At this point, r8 has argv, rcx has argc
-copy_args_to_stack:
+
         mov r10, rcx   // Store argc in r10
-        sub r10, 4     // Adjust for args that are passed through the registers
+        add r10, 3     // function, callInfo, ..., + 1 for alignment
         and r10, -2    // Mask off the lower bit to 16 byte align the stack
         shl r10, 3     // Calculate space for remaining args (* sizeof(Var*))
 
         cmp r10, 1000h // If the space is large, make sure the stack is committed
-        jl  allocate_stack
+        jl  LOCAL_LABEL(allocate_stack)
         // xplat-todo: Figure out if we need to implement __chkstk
         // call __chkstk
 
-allocate_stack:
+LOCAL_LABEL(allocate_stack):
         sub rsp, r10   // Allocate the stack space
-        sub r10, 8     // r10 = ((argc - 4) - 1) * 8
-        lea rax, [r8 + 20h] // rax = &argv[4]
+        mov qword ptr [rsp], rdi        // function
+        mov qword ptr [rsp + 8h], rsi   // callInfo
+        cmp rcx, 0
+        je LOCAL_LABEL(args_setup_done)
 
-        // Copying from rax+r10 to rsp+r10
-        // Can use rcx as temporary since argc is not needed anymore
-copy_args:
+        // At this point, r8 has argv, rcx has argc
+LOCAL_LABEL(copy_args_to_stack):
+        mov r10, rcx
+        dec r10
+        shl r10, 3              // r10 = (argc - 1) * 8
+        lea rax, [r8]           // rax = argv
+        lea rbx, [rsp + 10h]    // rbx = &rsp[2]
+
+        // Copying from rax+r10 to rbx+r10
+LOCAL_LABEL(copy_args):
         mov r12, qword ptr [rax + r10]
```

Comment:
Yes we can. Didn't try that for sake of less change; and most of the time we have only a handful of args.